### ISSUE TITLE
[IMPROVEMENT] Optimize Engine v2 Online Rebuilding Load

### ISSUE THREAD

**Issue by hookak** [longhorn/longhorn#9999]
## Is your improvement request related to a feature? Please describe (👍 if you like this request)

In Engine v2, when executing replicaAdd, a shallowCopy operation is performed. Even after the engine resumes, the shallowCopy load significantly impacts performance, leading to a significant reduction in Client I/O.

<img width="644" alt="image" src="https://github.com/user-attachments/assets/75f690c5-27df-4160-ad1b-72adb01cae58" />

## Describe the solution you'd like

I attempted to limit the load of shallowCopy by applying `bdev_set_qos_limit` to the source and destination bdev. However, the SPDK-QoS (bdev_set_qos_limit) configuration for the rebuilding bdev did not work as expected.

It is essential to provide admin the ability to limit rebuilding load and prioritize Client I/O during such operations.

For reference, the following bdev was used in the QoS configuration:
`sdc/pvc-87fdfadb-cdae-48d6-943a-5aa4c44df03a-r-c45ca8f3-rebuilding`
 ```
 go-spdk-helper bdev get
 ...
        "assigned_rate_limits": {
            "rw_ios_per_sec": 1000, // multiple of 1000
            "rw_mbytes_per_sec": 10,
            "r_mbytes_per_sec": 10,
            "w_mbytes_per_sec": 10
        },
...
 ```

Is there anything else that I needs further investigate? It would be great to see this included in a future feature.

**Comment by derekbit** [comment]
cc @DamiaSan @shuo-wu @PhanLe1010  @innobead

**Comment by DamiaSan** [comment]
> However, the SPDK-QoS (bdev_set_qos_limit) configuration for the rebuilding bdev did not work as expected.

Could you describe better the operations performed and the results obtained?

**Comment by hookak** [comment]
@DamiaSan 

The basic idea is to use SPDK's bdev_set_qos_limit. Since there is no specific function in the go-spdk-engine, I applied it using a Python script:

```
import json
import socket


sock_path = '/var/tmp/spdk.sock'

# bdev_set_qos_limit rpc requests
rpc_request = {
    "jsonrpc": "2.0",
    "method": "bdev_set_qos_limit",
    "params": {
        "name": "pvc-02703413-b541-4688-881a-a644b28ba039-e-0",  
        "rw_ios_per_sec": 1000,  
        "rw_mbytes_per_sec": 10,
        "r_mbytes_per_sec" : 10,
        "w_mbytes_per_sec" : 10
    },
    "id": 1
}

with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
    sock.connect(sock_path)
    sock.sendall(json.dumps(rpc_request).encode('utf-8'))
    
    
    response = sock.recv(4096)
    print("Response:", response.decode('utf-8'))
```

* It works fine when applying to engine bdev in params.name.
* However, it does not work when applying to rebuilding bdev in params.name.
  * Need to check SPDK whether QoS is not applied to shallow-copy 

As a result, I have not yet found an appropriate way to control the load of rebuilding during shallow copy.

**Comment by hookak** [comment]
> Do you execute the `bdev_set_qos_limit` before to start the shallow copy right?

No, I executed `bdev_set_qos_limit` while the shallow copy was already in progress. Since the shallow copy transfers data from the `src lvol` to the `target bdev`, I applied `bdev_set_qos_limit` to the target bdev (as QoS cannot be applied to lvols). Could this cause the `bdev_set_qos_limit` to not function correctly because it cannot be applied to the src lvol?


> For all SPDK commands that are not covered by go-spdk-helper, you can use rpc.py that is in the scripts directory of SPDK repo.

I am unable to use rpc.py because the SPDK module cannot be found. Could you guide me on how to use rpc.py within an lhim?

```
instance-manager-1eed0c33bee28248e5c22d78bf8c8e5a:/usr/src/spdk/scripts # /usr/src/spdk/scripts/rpc.py -h
Traceback (most recent call last):
  File "/usr/src/spdk/scripts/rpc.py", line 24, in <module>
    import spdk.rpc as rpc  # noqa
    ^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'spdk'
```

**Comment by derekbit** [comment]
Can you try 
```
zypper install git
https://github.com/longhorn/spdk.git
cd spdk
./scripts/rpc.py
```

There are some deps issues in `/usr/src/spdk/scripts/rpc.py`. Not time for fixing them :(

**Comment by DamiaSan** [comment]
> No, I executed bdev_set_qos_limit while the shallow copy was already in progress. Since the shallow copy transfers data from the src lvol to the target bdev, I applied bdev_set_qos_limit to the target bdev (as QoS cannot be applied to lvols). Could this cause the bdev_set_qos_limit to not function correctly because it cannot be applied to the src lvol?

No, youy are doing the right thing, i.e. applying the qos to the destination bdev. Looking at the code, the qos should be applied also if the shallow copy is already in progress. Shallow copy copy one cluster at a time, and every write operation check if qos is enabled or not, so it is weird you don't see any changes after having applied the qos.

**Comment by hookak** [comment]
To clarify, I apologize for the earlier confusion, but it seems the performance degradation of client I/O is actually due to `snapshots`, not shallow-copy as I initially thought.
When a snapshot is created, a 4K write operation can result in a worst-case scenario of {1M read + 1M write}, caused by Copy-on-Write (CoW) and the 1M cluster size.
The reason for the significantly low client I/O appears to be the severe I/O amplification issue at the actual device level.


I found the discussion you had about Copy-on-Write (CoW) during the design of online rebuilding (https://github.com/longhorn/longhorn/issues/5866#issuecomment-1630251832).

While snapshots are essential for supporting online rebuilding, the I/O amplification issue seems severe and likely needs improvement.
I’m curious to hear your thoughts on this. @DamiaSan

**Comment by hookak** [comment]
> Shallow copy copy one cluster at a time, and every write operation check if qos is enabled or not, so it is weird you don't see any changes after having applied the qos.

For reference, I tested bdev_set_qos_limit multiple times on the destination bdev during shallow-copy, and I have confirmed that it is now working correctly.
Thank you.

**Comment by hookak** [comment]
When adding a bdev to bdev_raid, SPDK rebuild allows volume addition without requiring a snapshot. Is there a specific reason why `replicaAdd` is implemented based on snapshots instead of using SPDK rebuild?

Considering the significant performance degradation caused by snapshots, it seems reasonable to avoid triggering a snapshot through the system’s operations(node evict, unexpected im shutdown) unless explicitly performed by the user.
@DamiaSan @shuo-wu

**Comment by DamiaSan** [comment]
> When adding a bdev to bdev_raid, SPDK rebuild allows volume addition without requiring a snapshot. Is there a specific reason why `replicaAdd` is implemented based on snapshots instead of using SPDK rebuild?
> 
> Considering the significant performance degradation caused by snapshots, it seems reasonable to avoid triggering a snapshot through the system’s operations(node evict, unexpected im shutdown) unless explicitly performed by the user. @DamiaSan @shuo-wu

SPDK raid rebuild functionality simply copy all the data of the base bdev, but with logical volumes we need to recreate the entire snapshot stack of the volume, writing data layer by layer.

**Comment by hookak** [comment]
I’m sorry, I wasn’t able to check your response earlier.

> SPDK raid rebuild functionality simply copy all the data of the base bdev, but with logical volumes we need to recreate the entire snapshot stack of the volume, writing data layer by layer.

That’s correct. However, compared to the traditional method, the total amount of copied data does not differ significantly.
The key point is that during an online rebuild, it is possible to configure the system so that client I/O is minimally impacted.

> It truly depends on the type of your I/Os, what kind of improvements would you require?

Indeed, most of our block storage workload consists of writes.
Therefore, our proposal focuses on optimizing for write workloads, particularly small random writes.

Above all, our ultimate goal is to minimize the impact on client I/O, regardless of the system events.


I’ve opened the same issue on https://github.com/longhorn/longhorn/issues/10163
It would be great to continue the discussion there.
If you’re okay with it, I’ll proceed to close the issue. @DamiaSan
